pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,367 @@
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"sync"
+	"sync/atomic"
+
+	"github.com/nspcc-dev/neo-go/pkg/config"
+	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	gio "github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/network/bqueue"
+	"github.com/nspcc-dev/neo-go/pkg/network/payload"
+	"github.com/nspcc-dev/neo-go/pkg/services/oracle/neofs"
+	"github.com/nspcc-dev/neo-go/pkg/wallet"
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+const (
+	ModeIndexSearch = "indexSearch"
+	ModeOidSearch   = "oidSearch"
+
+	oidSize   = sha256.Size
+	batchSize = 2000
+)
+
+// Ledger is an interface to Blockchain sufficient for Service.
+type Ledger interface {
+	AddBlock(block *block.Block) error
+	GetConfig() config.Blockchain
+	BlockHeight() uint32
+	AddHeaders(...*block.Header) error
+	HeaderHeight() uint32
+}
+
+// Service is a service that fetches blocks from NeoFS.
+type Service struct {
+	chain             Ledger
+	client            *client.Client
+	log               *zap.Logger
+	quit              chan bool
+	toExitCh          chan struct{}
+	cfg               config.NeoFSBlockFetcher
+	queue             *bqueue.Queue
+	account           *wallet.Account
+	started           atomic.Bool
+	stateRootInHeader bool
+}
+
+// New creates a new BlockFetcherService.
+func New(chain Ledger, cfg config.NeoFSBlockFetcher, queue *bqueue.Queue, logger *zap.Logger) *Service {
+	var (
+		account *wallet.Account
+	)
+	if cfg.UnlockWallet.Path != "" {
+		wallet, err := wallet.NewWalletFromFile(cfg.UnlockWallet.Path)
+		if err != nil {
+			logger.Error("Failed to load wallet from file", zap.Error(err))
+		}
+		for _, acc := range wallet.Accounts {
+			if err := acc.Decrypt(cfg.UnlockWallet.Password, wallet.Scrypt); err == nil {
+				account = acc
+				logger.Info("Successfully unlocked account from wallet")
+				break
+			}
+		}
+		if account == nil {
+			logger.Info("No valid account found in wallet, creating a new one")
+			err = wallet.CreateAccount("blockfetcher", cfg.UnlockWallet.Password)
+			if err != nil {
+				return nil
+			}
+			account = wallet.Accounts[len(wallet.Accounts)-1]
+		}
+	}
+
+	return &Service{
+		chain:             chain,
+		log:               logger,
+		quit:              make(chan bool),
+		toExitCh:          make(chan struct{}),
+		cfg:               cfg,
+		queue:             queue,
+		account:           account,
+		stateRootInHeader: chain.GetConfig().StateRootInHeader,
+	}
+}
+
+// Start implements the Service interface.
+func (bfs *Service) Start() {
+	if bfs.started.CompareAndSwap(false, true) {
+		bfs.log.Info("Starting Block Fetcher Service")
+		var err error
+		bfs.client, err = neofs.GetSDKClient(context.Background(), bfs.cfg.Addresses[0])
+		if err != nil {
+			bfs.log.Error("Failed to create SDK client", zap.Error(err))
+			close(bfs.quit)
+			return
+		}
+		defer bfs.client.Close()
```

Comment:
This `defer` is not correct here because client will be closed immediately after `Start` exit, and you run the main loop in a separate goroutine. The result is all requests will be failed. Have you tested this code, does it works?

Reply:
Close client in `Shutdown`.